### PR TITLE
IP-based validation

### PR DESCRIPTION
**Pull request by probakowski**
This change adds IP-based validation for TLS and SSH certificates.
There's new option in role definition:
```yaml
kind: role
metadata:
  name: dev
spec:
  options:
    pin_source_ip: true
```
When that is set to true client IP must be the same when generating certificates and using them. 
It applies to SSH (through `source_address` critical option that should be supported by both teleport and sshd), Kubernetes, Applications and Databases Access (through `(a *authorizer) Authorize` method).

Most of the lines changed are from regenerating `types.proto`, change itself is not that big

Relates https://github.com/gravitational/teleport/pull/11719

### CODE REVIEW

File: lib/auth/permissions.go [hunk 1]
```diff
@@ -178,6 +179,17 @@ func (a *authorizer) Authorize(ctx context.Context) (*Context, error) {
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+	if authContext.Checker.PinSourceIP() {
```
Comment (probakowski) [inline lib/auth/permissions.go:182]: This differs between SSH and TLS - SSH certificates validate IPs if `pin_source_ip` was set during certificate creation, TLS certificates validate IPs if `pin_source_ip` is set when user tries to use the certificate. We can modify it so both ways work the same by storing state `pin_source_ip` in custom extension and checking that in the line above.

File: api/types/session.go [hunk 2]
```diff
@@ -512,6 +512,8 @@ type NewWebSessionRequest struct {
 	LoginTime time.Time
 	// AccessRequests contains the UUIDs of the access requests currently in use.
 	AccessRequests []string
+	// ClientIP is an IP of the client requesting web session.
+	ClientIP string
```
Comment (marcoandredinis) [inline api/types/session.go:516]: I was expecting to see https://pkg.go.dev/net#IP here
Any reason to use string instead?

File: lib/auth/native/native.go [hunk 3]
```diff
@@ -264,6 +264,9 @@ func (k *Keygen) GenerateUserCertWithoutValidation(c services.UserCertParams) ([
 	}
 	if c.ClientIP != "" {
 		cert.Permissions.Extensions[teleport.CertExtensionClientIP] = c.ClientIP
+		if c.PinSourceIP {
+			cert.Permissions.CriticalOptions = map[string]string{"source-address": c.ClientIP}
```
Comment (marcoandredinis) [inline lib/auth/native/native.go:268]: I think we should add a bit more context of what `source-address` means/enforces
Maybe add a link to the OpenSSH docs? https://cvsweb.openbsd.org/src/usr.bin/ssh/PROTOCOL.certkeys?annotate=HEAD

I also noted that the docs says it requires CIDR notation
Would it work with just the IP? or would we need to add `/32` to turn this into a valid CIDR notation?
The golang's ssh server implementation supports both: https://github.com/golang/crypto/blob/2cf3adece1227c48e1673f1c37d70357e1a6b9d3/ssh/server.go#L306-L319
I'm not sure about the OpenSSH implementation https://github.com/openssh/openssh-portable/blob/2dc328023f60212cd29504fc05d849133ae47355/addrmatch.c#L109

File: lib/auth/apiserver.go [hunk 4]
```diff
@@ -758,6 +759,7 @@ func (s *APIServer) authenticateWebUser(auth ClientI, w http.ResponseWriter, r *
 		return nil, trace.Wrap(err)
 	}
 	req.Username = p.ByName("user")
+	req.ClientIP = strings.Split(r.RemoteAddr, ":")[0]
```
Comment (marcoandredinis) [inline lib/auth/apiserver.go:762]: Do you think we should add some checks here?
Hopefully we always have "IP:port" here, but better be safe than panic

File: api/types/session.go (hunk 2 above)
Comment (probakowski) [reply api/types/session.go:516]: We store client IP in string format in different places (like `UserCertParams`, `Identity` etc) already, I just kept it consistent

File: lib/auth/native/native.go (hunk 3 above)
Comment (probakowski) [reply lib/auth/native/native.go]: I've created const here and added a comment, hopefully it's clearer now.
Good catch with CIDR, I've added `/32` so we are always safe.

File: lib/auth/apiserver.go (hunk 4 above)
Comment (probakowski) [reply lib/auth/apiserver.go]: It wouldn't panic, splitting is always safe, it'd simply fail (with some access denied exception) down the road.
Either way, I've switched to [utils.Host](https://github.com/gravitational/teleport/blob/master/lib/utils/utils.go#L282) as @rosstimothy suggested

File: lib/auth/apiserver.go [hunk 5]
```diff
@@ -771,6 +773,7 @@ func (s *APIServer) authenticateSSHUser(auth ClientI, w http.ResponseWriter, r *
...
```
[the remark this answers is not in the document]
Comment (probakowski) [reply lib/auth/apiserver.go]: Yep, switched to [utils.Host](https://github.com/gravitational/teleport/blob/master/lib/utils/utils.go#L282)

File: lib/auth/permissions.go [hunk 6]
```diff
@@ -178,6 +179,17 @@ func (a *authorizer) Authorize(ctx context.Context) (*Context, error) {
...
+		value := ctx.Value(ContextClientAddr)
+		addr, ok := value.(*net.TCPAddr)
+		if !ok {
+			return nil, trace.AccessDenied("client IP can't be checked but is required")
+		}
+		pinnedIP := authContext.Identity.GetIdentity().ClientIP
+		if pinnedIP != addr.IP.String() {
+			return nil, trace.AccessDenied("client IP %s does not match the expected %s", addr.IP, pinnedIP)
```
[the remark this answers is not in the document]
Comment (probakowski) [reply lib/auth/permissions.go]: changed

File: lib/multiplexer/testproxy.go [hunk 7]
```diff
@@ -118,13 +120,20 @@ func (p *TestProxy) sendProxyLine(clientConn, serverConn net.Conn) error {
 	if err != nil {
 		return trace.Wrap(err)
 	}
+	sourceAddr := clientAddr
+	if p.sourceAddr != "" {
+		sourceAddr, err = utils.ParseAddr(p.sourceAddr)
+		if err != nil {
+			return trace.Wrap(err)
+		}
```
[the remark this answers is not in the document]
Comment (probakowski) [reply lib/multiplexer/testproxy.go]: That makes sense, I've moved it

File: lib/auth/permissions.go [hunk 8]
```diff
@@ -178,6 +179,17 @@ func (a *authorizer) Authorize(ctx context.Context) (*Context, error) {
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+	if authContext.Checker.PinSourceIP() {
+		value := ctx.Value(ContextClientAddr)
+		addr, ok := value.(*net.TCPAddr)
+		if !ok {
+			return nil, trace.AccessDenied("client IP can't be checked but is required")
```
[the remark this answers is not in the document]
Comment (probakowski) [reply lib/auth/permissions.go]: changed

File: api/types/session.go (hunk 2 above)
Comment (marcoandredinis) [reply api/types/session.go:516]: Thank you for clarifying :+1:

File: integration/db_integration_test.go [hunk 9]
```diff
@@ -53,6 +54,78 @@ import (
...
+		Username:   pack.root.user.GetName(),
+		RouteToDatabase: tlsca.RouteToDatabase{
+			ServiceName: pack.root.postgresService.Name,
+			Protocol:    pack.root.postgresService.Protocol,
+			Username:    "postgres",
+			Database:    "test",
+		},
+	})
+	require.Error(t, err)
+	require.Contains(t, err.Error(), "client IP 1.2.3.4 does not match the expected 127.0.0.1")
```
Comment (marcoandredinis) [inline integration/db_integration_test.go:126]:
```suggestion
	require.Contains(t, err.Error(), "client IP 1.2.3.4 does not match the pinned 127.0.0.1")
```

**Review by marcoandredinis — APPROVED** [review]
LGTM

**Comment by codingllama** [comment]
Best PR number ever.

File: lib/auth/permissions.go [hunk 10]
```diff
@@ -178,6 +179,16 @@ func (a *authorizer) Authorize(ctx context.Context) (*Context, error) {
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
+	if authContext.Checker.PinSourceIP() {
+		addr, ok := ctx.Value(ContextClientAddr).(*net.TCPAddr)
+		if !ok {
+			return nil, trace.AccessDenied("IP pinning enabled but client IP is missing")
+		}
+		pinnedIP := authContext.Identity.GetIdentity().ClientIP
+		if pinnedIP != "" && pinnedIP != addr.IP.String() {
```
[the remark this answers is not in the document]
Comment (probakowski) [reply lib/auth/permissions.go:188]: It shows in UI (probably in other parts as well) when you're logged in before upgrade (i.e. running from `master`) and then refresh page after upgrade (so running from this branch)  with pinning enabled. It results in access denied error with no way to log out.

**Review by codingllama — COMMENTED** [review]
It looks like you are covered in terms of reviewers, so I just had a quick glance. Seems reasonable to me.

**Review by rosstimothy — APPROVED** [review]

File: lib/auth/native/native.go [hunk 11]
```diff
@@ -264,6 +269,11 @@ func (k *Keygen) GenerateUserCertWithoutValidation(c services.UserCertParams) ([
 	}
 	if c.ClientIP != "" {
 		cert.Permissions.Extensions[teleport.CertExtensionClientIP] = c.ClientIP
+		if c.PinSourceIP {
+			// exact IP in CIDR notation, all bits matter
+			cidr := c.ClientIP + "/32"
```
Comment (reedloden) [inline lib/auth/native/native.go:274]: What about IPv6? This feels very IPv4-only.

**Review by strideynet — COMMENTED** [review]
I pulled this PR down locally and built it to give it a go on my EC2 instance (fronted by an NLB). Unfortunately, it doesn't look like the correct IP addresses are being detected. 

Here is an example of a certificate generated when I log into `tsh`:
```
Certificate Type: ssh-rsa-cert-v01@openssh.com
Public Key:       SHA256:/rtWNdTzAet7biqx/Ag9AEIjFBUK+k58k+6kMSoIhG0
Signing CA:       SHA256:jx4pL7noGJ2U+qo6+vwDMXXu6mhTD5EwnGPKFzhGxMQ
Key ID:           noahstride
Principals:       root,ubuntu,ec2-user,-teleport-internal-join
Valid After:      May 20 09:11:26 UTC
Valid Before:     May 20 21:12:26 UTC
Critical Options: none
Extensions:
    teleport-traits: {"aws_role_arns":null,"db_names":null,"db_users":null,"kubernetes_groups":null,"kubernetes_users":null,"logins":["root","ubuntu","ec2-user"],"windows_logins":null}
    client-ip:       127.0.0.1
    permit-agent-forwarding
    permit-port-forwarding
    permit-pty
    teleport-roles:            {"version":"v1","roles":["editor","access"]}
    teleport-route-to-cluster: noah-test.teleportdemo.net
```

I've uploaded a gist of my configuration to https://gist.github.com/strideynet/4e3c178aa5844fec47eee5d17de2c4a4 . I wouldn't be surprised if I have something misconfigured, but I've tried a few things to try and get this working.

I added some additional logging in the proxy webapi and the auth webapi to dump `r.RemoteAddr()`, yielding the following results (my IP has been redacted, but the value was correct):
```
2022-05-20T11:23:45Z INFO [WEB]       mfaLoginFinish 123.456.123.456:63011 web/apiserver.go:1793
2022-05-20T11:23:45Z INFO [AUTH]      authenticateSSHUser 127.0.0.1:47010 auth/apiserver.go:793
```

**Comment by r0mant** [comment]
@probakowski Are we closing this in favor of https://github.com/gravitational/teleport/pull/13243?

**Comment by probakowski** [comment]
Yes, this PR had some wrong assumptions and didn't work as expected. I'll close it in favor of https://github.com/gravitational/teleport/pull/13243

**Pull request closed without merging**